src/dpdkstat.c

```diff
@@ -0,0 +1,775 @@
+     g_configuration->helper_status == DPDK_HELPER_GRACEFUL_QUIT) {
+      int action = DPDK_HELPER_ACTION_SEND_STATS;
+      if(g_configuration->num_xstats == 0)
+        action = DPDK_HELPER_ACTION_COUNT_STATS;
+      /* Spawn the helper thread to count stats or to read stats. */
+      int err = dpdk_helper_spawn(action);
+      if (err) {
+        ERROR("dpdkstat: error spawning helper %s\n", strerror(errno));
+        return -1;
+      }
+    }
+
+  int exit_status;
+  pid_t ws = waitpid(g_configuration->helper_pid, &exit_status, WNOHANG);
+  /*
+   * Conditions under which to respawn helper:
+   *  waitpid() fails, helper process died (or quit), so respawn
+   */
+  int respawn_helper = 0;
+  if (ws != 0) {
+    respawn_helper = 1;
+  }
+
+  char buf[DPDKSTAT_MAX_BUFFER_SIZE];
+  char out[DPDKSTAT_MAX_BUFFER_SIZE];
+
+  /* non blocking check on helper logging pipe */
+  struct pollfd fds;
+  fds.fd = g_configuration->helper_pipes[0];
+  fds.events = POLLIN;
+  int data_avail = poll(&fds, 1, 0);
+  while (data_avail) {
+    int nbytes = read(g_configuration->helper_pipes[0], buf, sizeof(buf));
+    if (nbytes <= 0)
+      break;
+    ssnprintf( out, nbytes, "%s", buf);
+    DEBUG("dpdkstat: helper-proc: %s\n", out);
+  }
+
+  if (respawn_helper) {
+    if (g_configuration->helper_pid)
+      dpdk_helper_exit(RESET);
+    dpdk_helper_spawn(DPDK_HELPER_ACTION_COUNT_STATS);
+  }
+
+  struct timespec helper_kick_time;
+  clock_gettime(CLOCK_REALTIME, &helper_kick_time);
+  /* Kick helper process through SHM */
+  sem_post(&g_configuration->sema_helper_get_stats);
+
+  struct timespec ts;
+  cdtime_t now = cdtime();
+  CDTIME_T_TO_TIMESPEC(now + g_configuration->interval, &ts);
+  ret = sem_timedwait(&g_configuration->sema_stats_in_shm, &ts);
+  if (ret == -1 && errno == ETIMEDOUT) {
+    DEBUG("dpdkstat: timeout in collectd thread: is a DPDK Primary running? \n");
+    return 0;
+  }
+
+  /* Dispatch the stats.*/
+  int count = 0, port_num = 0;
+
+  for (uint32_t i = 0; i < g_configuration->num_ports; i++) {
+    cdtime_t time = g_configuration->port_read_time[i];
+    char dev_name[64];
+    int len = g_configuration->num_stats_in_port[i];
+
+    while(!(g_configuration->enabled_port_mask & (1 << port_num)))
+      port_num++;
+
+    if (g_configuration->port_name[i][0] != 0)
+      ssnprintf(dev_name, sizeof(dev_name), "%s", g_configuration->port_name[i]);
+    else
+      ssnprintf(dev_name, sizeof(dev_name), "port.%d", port_num);
+    struct rte_eth_xstats *xstats = (&g_configuration->xstats);
+    xstats += count; /* pointer arithmetic to jump to each stats struct */
+    for (int j = 0; j < len; j++) {
+      value_t dpdkstat_values[1];
+      value_list_t dpdkstat_vl = VALUE_LIST_INIT;
+      char *type_end;
+
+      dpdkstat_values[0].derive = (derive_t) xstats[j].value;
+      dpdkstat_vl.values = dpdkstat_values;
+      dpdkstat_vl.values_len = 1; /* Submit stats one at a time */
+      dpdkstat_vl.time = time;
```

Comment:
Please use `g_configuration->port_read_time[i]` directly here (or rename `time` to `port_read_time`).